### PR TITLE
Xml operations

### PR DESCRIPTION
**Pull request by tethryus**
Enhancements, Fixes, xml files now supported

Signed off by: Sorin Moldovan sorin.moldovan@hpe.com

### CODE REVIEW

**Comment by mihaitusa** [comment]
@tethryus  Please let just one of the labels... is either "ready for merge" or "in progress". Can we review it? Is ready?

**Comment by tethryus** [comment]
@mihaitusa Yes, it can be reviewed, please do, whenever you have the time.

**Comment by mihaitusa** [comment]
@tethryus  master @gituser173 will review it

File: content/io/cloudslang/base/xml/add_attribute.sl [hunk 1]
```diff
@@ -9,7 +9,10 @@
 #!!
 #! @description: Adds an attribute to an XML element or replaces the value if the attribute already exists.
 #!
-#! @input xml_document: XML string in which to add attribute
+#! @input xml_document: XML string or file in which to add attribute
+#! @input xml_document_source: xml document type
+#!                             Default: xmlStirng
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/add_attribute.sl]: typo

File: content/io/cloudslang/base/xml/add_attribute.sl [hunk 2]
```diff
@@ -35,6 +39,11 @@ operation:
     - xmlDocument:
         default: ${get("xml_document", "")}
         private: true
+    - xml_document_source:
+        default: xmlString
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/add_attribute.sl]: ?

File: content/io/cloudslang/base/xml/append_child.sl [hunk 3]
```diff
@@ -9,7 +9,10 @@
 #!!
 #! @description: Appends a child to an XML element.
 #!
-#! @input xml_document: XML string to append a child in
+#! @input xml_document: XML string or file to append a child in
+#! @input xml_document_source: xml document type
+#!                             Default: xmlStirng
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/append_child.sl]: typo

File: content/io/cloudslang/base/xml/append_child.sl [hunk 4]
```diff
@@ -34,6 +38,11 @@ operation:
     - xmlDocument:
         default: ${get("xml_document", "")}
         private: true
+    - xml_document_source:
+        default: xmlString
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/append_child.sl]: same question as above

File: content/io/cloudslang/base/xml/edit_xml.sl [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+#
+########################################################################################################################
+#!!
+# @description: Edits an XML document.
+#
+# @input xml: The XML (in the form of a String).
+# @input file_path: absolute or remote path of the XML file.
+# @input action: the edit action to take place.
+#                Valid values: delete, insert, append, subnode, move, rename, update.
+# @input xpath_1: the XPath Query that is wanted to be run. The changes take place at the resulting elements.
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/edit_xml.sl]: the XPath Query to be run

File: content/io/cloudslang/base/xml/edit_xml.sl [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+#                          the parser will not include external general entities.
+#                          When the "http://xml.org/sax/features/external-parameter-entities" feature is set to "false"
+#                          the parser will not include external parameter entities or the external DTD subset.
+#                          If any of the validations fails, the operation will fail with an error message describing the problem.
+#                          Default value:
+#                          http://apache.org/xml/features/disallow-doctype-decl true
+#                          http://xml.org/sax/features/external-general-entities false
+#                          http://xml.org/sax/features/external-parameter-entities false
+#
+# Notes:
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/edit_xml.sl]: maybe move the notes at the end of the description, WDYT?

File: content/io/cloudslang/base/xml/edit_xml.sl [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+operation:
+  name: edit_xml
+
+  inputs:
+    - xml:
+        required: false
+    - file_path:
+        required: false
+    - filePath:
+        default: ${get("file_path", "")}
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/edit_xml.sl]: add `required: false` here because filePath could have "" value and "" is not accepted for required inputs

File: content/io/cloudslang/base/xml/edit_xml.sl [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        private: true
+    - action
+    - xpath_1
+    - xpath1:
+        default: ${get("xpath_1", "")}
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/edit_xml.sl]: same here

File: content/io/cloudslang/base/xml/insert_before.sl [hunk 9]
```diff
@@ -9,7 +9,10 @@
 #!!
 #! @description: Appends a child to an XML element.
 #!
-#! @input xml_document: XML string to insert element in
+#! @input xml_document: XML string or file to insert element in
+#! @input xml_document_source: xml document type
+#!                             Default: xmlStirng
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/insert_before.sl]: typo, again

File: content/io/cloudslang/base/xml/insert_before.sl [hunk 10]
```diff
@@ -34,6 +38,11 @@ operation:
     - xmlDocument:
         default: ${get("xml_document", "")}
         private: true
+    - xml_document_source:
+        default: xmlString
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/insert_before.sl]: same question, what is `xmlString`?

File: content/io/cloudslang/base/xml/remove.sl [hunk 11]
```diff
@@ -37,6 +41,11 @@ operation:
     - xmlDocument:
         default: ${get("xml_document", "")}
         private: true
+    - xml_document_source:
+        default: xmlString
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/remove.sl]: see above

File: content/io/cloudslang/base/xml/remove.sl [hunk 12]
```diff
@@ -37,6 +41,11 @@ operation:
     - xmlDocument:
         default: ${get("xml_document", "")}
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/remove.sl]: `required:false` here

File: content/io/cloudslang/base/xml/validate.sl [hunk 13]
```diff
@@ -9,11 +9,44 @@
...
-#! @input xml_document: XML string to test
+#! @input xml_document: XML string, file or url to test
+#! @input xml_document_source: xml document type
+#!                             Default: xmlStirng
+#!                             Accepted values: xmlString, xmlPath, xmlUrl
 #! @input xsd_document: XSD to test given XML against
 #!                      optional
-#! @output return_result: exception in case of failure, success message otherwise
-#! @output result_text: 'success' or 'failure'
+#! @input xsd_document_source: xml document type
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/validate.sl]: isn't this duplicated?

File: test/io/cloudslang/base/xml/edit_xml.inputs.yaml [hunk 14]
```diff
@@ -0,0 +1,19 @@
+#   (c) Copyright 2016 Hewlett-Packard Enterprise Development Company, L.P.
+#   All rights reserved. This program and the accompanying materials
+#   are made available under the terms of the Apache License v2.0 which accompany this distribution.
+#
+#   The Apache License is available at
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+####################################################
+testEditXmlAppendChildSuccess:
```
Comment (gituser173) [inline test/io/cloudslang/base/xml/edit_xml.inputs.yaml]: add more tests

File: content/io/cloudslang/base/xml/remove.sl (hunk 12 above)
Comment (tethryus) [inline content/io/cloudslang/base/xml/remove.sl]: addded

File: content/io/cloudslang/base/xml/append_child.sl (hunk 3 above)
Comment (tethryus) [inline content/io/cloudslang/base/xml/append_child.sl]: fixed

File: content/io/cloudslang/base/xml/add_attribute.sl (hunk 1 above)
Comment (tethryus) [inline content/io/cloudslang/base/xml/add_attribute.sl]: fixed

File: content/io/cloudslang/base/xml/add_attribute.sl (hunk 2 above)
Comment (tethryus) [inline content/io/cloudslang/base/xml/add_attribute.sl]: as a string

File: content/io/cloudslang/base/xml/append_child.sl (hunk 4 above)
Comment (tethryus) [inline content/io/cloudslang/base/xml/append_child.sl]: as a string

File: content/io/cloudslang/base/xml/edit_xml.sl (hunk 5 above)
Comment (tethryus) [inline content/io/cloudslang/base/xml/edit_xml.sl]: modified

File: content/io/cloudslang/base/xml/edit_xml.sl (hunk 6 above)
Comment (tethryus) [inline content/io/cloudslang/base/xml/edit_xml.sl]: notes and example moved to the end of the description, seems a better idea. thank you for pointing that out.

File: content/io/cloudslang/base/xml/edit_xml.sl (hunk 7 above)
Comment (tethryus) [inline content/io/cloudslang/base/xml/edit_xml.sl]: added

File: content/io/cloudslang/base/xml/edit_xml.sl (hunk 8 above)
Comment (tethryus) [inline content/io/cloudslang/base/xml/edit_xml.sl]: added in all places needed

File: content/io/cloudslang/base/xml/insert_before.sl (hunk 9 above)
Comment (tethryus) [inline content/io/cloudslang/base/xml/insert_before.sl]: corrected typo

File: content/io/cloudslang/base/xml/insert_before.sl (hunk 10 above)
Comment (tethryus) [inline content/io/cloudslang/base/xml/insert_before.sl]: as a string

File: content/io/cloudslang/base/xml/remove.sl (hunk 11 above)
Comment (tethryus) [inline content/io/cloudslang/base/xml/remove.sl]: as a stirng

File: content/io/cloudslang/base/xml/add_attribute.sl [hunk 15]
```diff
@@ -35,31 +38,41 @@ operation:
     - xmlDocument:
         default: ${get("xml_document", "")}
         private: true
+    - xml_document_source:
+        required: false
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/add_attribute.sl]: why did you add `required: false`? that is only needed in cases where you do something like:
`- xPathElementQuery:
         default: ${get("xpath_element_query", "")}`

File: content/io/cloudslang/base/xml/add_attribute.sl [hunk 16]
```diff
@@ -35,31 +38,41 @@ operation:
...
+        default: 'xmlString'
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/add_attribute.sl]: also why did you specify the default here if it is already specified in the get function of the default from xmlDocumentSource?

File: content/io/cloudslang/base/xml/add_attribute.sl [hunk 17]
```diff
@@ -35,31 +38,41 @@ operation:
...
     - attributeName:
         default: ${get("attribute_name", "")}
+        required: false
         private: true
     - value
     - secure_processing:
         required: false
     - secureProcessing:
         default: ${get("secure_processing", "false")}
+        required: false
```
Comment (gituser173) [inline content/io/cloudslang/base/xml/add_attribute.sl]: `required: false` not needed on this case

File: content/io/cloudslang/base/xml/edit_xml.sl (hunk 8 above)
Comment (gituser173) [inline content/io/cloudslang/base/xml/edit_xml.sl]: actually you added this even in places where it is not needed, like this one:
`      - secureProcessing:
         default: ${get("secure_processing", "false")}
+        required: false
         private: true`

**Comment by mihaitusa** [comment]
@tethryus  Update the branch and solve the conflicts.
